executor/executor.go

```diff
@@ -0,0 +1,602 @@
+		}
+		e.iter, _, err = e.scan.idx.Seek(txn, []interface{}{seekVal})
+		if err != nil {
+			return nil, types.EOFAsNil(err)
+		}
+	}
+	for {
+		if e.finished {
+			return nil, nil
+		}
+		idxKey, h, err := e.iter.Next()
+		if err != nil {
+			return nil, types.EOFAsNil(err)
+		}
+		val := idxKey[0]
+		if !e.skipLowCmp {
+			cmp := indexCompare(val, e.lowVal)
+			if cmp < 0 || (cmp == 0 && e.lowExclude) {
+				continue
+			}
+			e.skipLowCmp = true
+		}
+		cmp := indexCompare(val, e.highVal)
+		if cmp > 0 || (cmp == 0 && e.highExclude) {
+			// This span has finished iteration.
+			e.finished = true
+			e.skipLowCmp = false
+			continue
+		}
+		var row *Row
+		row, err = e.lookupRow(h)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return row, nil
+	}
+}
+
+// comparison function that takes minNotNullVal and maxVal into account.
+func indexCompare(a interface{}, b interface{}) int {
+	if a == nil && b == nil {
+		return 0
+	} else if b == nil {
+		return 1
+	} else if b == nil {
+		return -1
+	}
+
+	// a and b both not nil
+	if a == plan.MinNotNullVal && b == plan.MinNotNullVal {
+		return 0
+	} else if b == plan.MinNotNullVal {
+		return 1
+	} else if a == plan.MinNotNullVal {
+		return -1
+	}
+
+	// a and b both not min value
+	if a == plan.MaxVal && b == plan.MaxVal {
+		return 0
+	} else if a == plan.MaxVal {
+		return 1
+	} else if b == plan.MaxVal {
+		return -1
+	}
+
+	n, err := types.Compare(a, b)
+	if err != nil {
+		// Old compare panics if err, so here we do the same thing now.
+		// TODO: return err instead of panic.
+		panic(fmt.Sprintf("should never happend %v", err))
+	}
+	return n
+}
+
+func (e *IndexRangeExec) lookupRow(h int64) (*Row, error) {
+	row := &Row{}
+	var err error
+	row.Data, err = e.scan.tbl.Row(e.scan.ctx, h)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	rowKey := &RowKeyEntry{
+		Tbl: e.scan.tbl,
+		Key: string(e.scan.tbl.RecordKey(h, nil)),
+	}
+	row.RowKeys = append(row.RowKeys, rowKey)
+	return row, nil
+}
+
+// Close implements Executor Close interface.
+func (e *IndexRangeExec) Close() error {
+	if e.iter != nil {
+		e.iter.Close()
+		e.iter = nil
+	}
+	e.finished = false
+	e.skipLowCmp = false
+	return nil
+}
+
+// IndexScanExec represents an index scan executor.
+type IndexScanExec struct {
+	tbl       table.Table
+	idx       kv.Index
+	fields    []*ast.ResultField
+	Ranges    []*IndexRangeExec
+	Desc      bool
+	rangeIdx  int
+	ctx       context.Context
+	valueType types.FieldType
+}
+
+// Fields implements Executor Fields interface.
+func (e *IndexScanExec) Fields() []*ast.ResultField {
+	return e.fields
+}
+
+// Next implements Executor Next interface.
+func (e *IndexScanExec) Next() (*Row, error) {
+	for e.rangeIdx < len(e.Ranges) {
+		ran := e.Ranges[e.rangeIdx]
+		row, err := ran.Next()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if row != nil {
+			for i, val := range row.Data {
+				e.fields[i].Expr.SetValue(val)
+			}
+			return row, nil
+		}
+		ran.Close()
```

Comment:
What is the cost of Close function?I see many places using xx.Close() in Next function.